conan/internal/model/conanfile_interface.py

```diff
@@ -18,6 +18,8 @@ def __eq__(self, other):
         The conanfile is a different entity per node, and conanfile equality is identity
         :type other: ConanFileInterface
         """
+        if self._conanfile.pref is not None and other._conanfile.pref is not None:
```

Comment:
This approach would ensure that conanfiles that share a pref are treated as the same, even if they are not the same object, which also solves the issue. Whichever solution is picked, the goal is to have the conanfile be identical for the same prefs in the deps graph so that `Node.propagate_downstream` does not get confused